src/Model/BSONDocument.php

```diff
@@ -32,6 +32,18 @@
  */
 class BSONDocument extends ArrayObject implements JsonSerializable, Serializable, Unserializable
 {
+    /**
+     * Deep clone this BSONDocument.
+     */
+    public function __clone()
+    {
+        foreach ($this as $key => $value) {
```

Comment:
This approach works if all arrays are BSONArray, but we previously talked about needing to copy elements within a basic array as well. That is possible if the user is only returning root and embedded documents as BSONDocument and requesting BSON arrays as basic PHP arrays.

Additionally, I believe this approach can segfault:

```php

class Foo {
    function __clone() {
        foreach ($this as $key => $value) {
            if (is_object($this->$key)) {
                $this->$key = clone $value;
            }
        }
    }
}

$a = new Foo;
$a->a = $a;
$b = clone $a;
```

It may be worth looking into how [DeepCopy](https://github.com/myclabs/DeepCopy) handles this by maintaining a hash of clones created during the top-most clone operation. This allows it to avoid a recursive rabbit hole. The readme explains some of the challenges with naive cloning approaches.

Looping in @derickr for advise here. Maybe we don't care about segfaulting on a recursive loop (PHP does that by design). I don't believe it's possible with a document originating directly from the server. As in my example, it requires purposefully creating a cycle in the object hierarchy.

Reply:
As @jmikola says, this can't happen with documents coming form the server, so I'm fine if we don't mitigate this issue.